### PR TITLE
[WC-2405]: Add barcode format controls to prevent incorrect scanning

### PR DESCRIPTION
**Pull request by HedwigAR**
### Pull request type

Bug fix (non-breaking change which fixes an issue)

---

### Description
We found a case where a Code-39 code was recognized as a RSS-14 code instead. Since we support both of them and they do look very similar, we've added an option to be able to specify what formats should be recognized.
Now the user can select if they want to be able to recognize all formats (this is the option that's enabled by default). If the user does not want this, they'll be able to specify only the codes they're expecting.
This PR also includes a fix for scanning with the mask enabled, where it wasn't scanning the full resolution.

### What should be covered while testing?
Testing should check whether it's still possible to recognize all barcode types when `useAllFormats` is set to Yes. Test should also check whether selecting specific formats is effective, and if this does exclude other codes from being recognized (aka, when only code-39 is selected it should no longer trigger on QR codes)

### CODE REVIEW

File: packages/pluggableWidgets/barcode-scanner-web/src/BarcodeScanner.editorConfig.ts [hunk 1]
```diff
@@ -13,6 +13,9 @@ export function getProperties(
     if (platform === "web") {
         transformGroupsIntoTabs(defaultProperties);
     }
+    if (_.useAllFormats) {
```
Comment (iobuhov) [inline packages/pluggableWidgets/barcode-scanner-web/src/BarcodeScanner.editorConfig.ts:16]: The `_` means "unused". But now I see that we use this param, can we assign a full name here?

**Review by iobuhov — CHANGES_REQUESTED** [review]

**Comment by iobuhov** [comment]
Great PR and description BTW 👍

File: packages/pluggableWidgets/barcode-scanner-web/src/BarcodeScanner.editorConfig.ts (hunk 1 above)
Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/BarcodeScanner.editorConfig.ts:16]: Sure!

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts [hunk 2]
```diff
@@ -83,6 +77,29 @@ export const useReader: UseReaderHook = args => {
...
+                BarcodeFormat.RSS_14,
+                BarcodeFormat.QR_CODE,
+                BarcodeFormat.DATA_MATRIX,
+                BarcodeFormat.AZTEC,
+                BarcodeFormat.PDF_417
+            ];
+        } else {
+            args.barcodeFormats.forEach((val, index) => {
+                formats[index] = BarcodeFormat[val.barcodeFormat];
+            });
```
[the remark this answers is not in the document]
Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts]: Ofcourse!

File: packages/pluggableWidgets/barcode-scanner-web/src/BarcodeScanner.editorConfig.ts [hunk 3]
```diff
@@ -1,18 +1,21 @@
...
 export function getProperties(
-    _: BarcodeScannerContainerProps,
+    barcodeScannerContainerProps: BarcodeScannerContainerProps,
     defaultProperties: Properties,
     platform: "web" | "desktop"
 ): Properties {
     if (platform === "web") {
         transformGroupsIntoTabs(defaultProperties);
     }
+    if (barcodeScannerContainerProps.useAllFormats) {
```
Comment (iobuhov) [inline packages/pluggableWidgets/barcode-scanner-web/src/BarcodeScanner.editorConfig.ts:16]: NIT: in other `editorConfig` files we use `values` as name

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts [hunk 4]
```diff
@@ -83,6 +77,27 @@ export const useReader: UseReaderHook = args => {
...
+                BarcodeFormat.CODE_128,
+                BarcodeFormat.ITF,
+                BarcodeFormat.RSS_14,
+                BarcodeFormat.QR_CODE,
+                BarcodeFormat.DATA_MATRIX,
+                BarcodeFormat.AZTEC,
+                BarcodeFormat.PDF_417
+            ];
+        } else {
+            args.barcodeFormats!.map((val, index) => (formats[index] = BarcodeFormat[val.barcodeFormat]));
```
Comment (iobuhov) [inline packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts:98]: nit:
```suggestion
            formats = args.barcodeFormats!.map(val => BarcodeFormat[val.barcodeFormat]);
```

Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts]: amazing

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts
Comment (iobuhov) [inline packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts:1]: Just one last thing: can you please fix eslint warnings in this file?

**Review by iobuhov — CHANGES_REQUESTED** [review]

**Review by iobuhov — APPROVED** [review]

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts [hunk 5]
```diff
@@ -8,63 +8,70 @@ import {
...
+        captureCanvas.width = videoCropWidth;
+        captureCanvas.height = videoCropWidth;
+
+        if (widthHeigher) {
+            videoCropWidth = videoCropWidth / aspectRatioWidth;
+        } else {
+            videoCropHeight = videoCropWidth;
+        }
+
+        const loop = (resolve: (value: Result) => void, reject: (reason?: Error) => void): void => {
```
Comment (gjulivan) [inline packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts:61]: function inside a function, either take it out or useCallback?

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts [hunk 6]
```diff
@@ -8,63 +8,70 @@ import {
...
+
         const captureCanvas = reader.createCaptureCanvas(videoRef.current!);
-        captureCanvas.width = cropWidth;
-        captureCanvas.height = cropWidth;
-        const loop = (resolve: (value: Result) => void, reject: (reason?: Error) => void) => {
+        captureCanvas.width = videoCropWidth;
+        captureCanvas.height = videoCropWidth;
+
+        if (widthHeigher) {
+            videoCropWidth = videoCropWidth / aspectRatioWidth;
```
Comment (gjulivan) [inline packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts:56]: if the value here is re-set, isn't the captureCanvas.width and height need to be adjusted again or no?

Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts:56]: That's a good point, I thought the drawImage would do that automatically but I'm not actually sure. I'll change it

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts (hunk 5 above)
Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts]: I took it out

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts [hunk 7]
```diff
@@ -8,89 +8,124 @@ import {
...
+                setTimeout(() => loop(resolve, reject, captureCanvas, videoCropWidth, videoCropHeight), 50);
+            } else {
+                reject(error);
+            }
+        }
+    };
+
     const scanWithCropOnce = (reader: BrowserMultiFormatReader): Promise<Result> => {
-        const cropWidth = videoRef.current!.videoWidth * scale;
+        const element = document.getElementById("canvas-middle-middle")!;
```
Comment (iobuhov) [inline packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts:79]: Maybe we should use refs? It's more react way.

Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts]: Thanks! I fixed it and used a ref instead, let me know if I did that right

File: packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts [hunk 8]
```diff
@@ -8,89 +8,124 @@ import {
...
+        videoCropWidth: number,
+        videoCropHeight: number
+    ): void => {
+        try {
+            const canvasContext = captureCanvas.getContext("2d", { willReadFrequently: true });
+            if (canvasContext !== null) {
+                canvasContext.drawImage(
+                    videoRef.current!,
+                    (videoRef.current!.videoWidth - videoCropWidth) / 2,
+                    (videoRef.current!.videoHeight - videoCropHeight) / 2,
```
[the remark this answers is not in the document]
Comment (HedwigAR) [reply packages/pluggableWidgets/barcode-scanner-web/src/hooks/useReader.ts]: You're right, I got lazy. Should be fixed now

**Review by iobuhov — APPROVED** [review]
LGTM

**Review by leonardomendix — APPROVED** [review]

**Review by iobuhov — APPROVED** [review]

**Pull request merged**